### DISCUSSION TITLE
Issue LangChain error: TypeError: Failed to fetch

### DISCUSSION THREAD

**Discussion by borthwick** [logancyang/obsidian-copilot#276] [Q&A]
I installed and configured copilot plugin on Mac m2, using LM Studio and "mistral instruct vO 1 7B Q4_K_M gguf"  

When I configure the plugin with QA as described in the documentation, I get the error ".LangChain error: TypeError: Failed to fetch  ".  

![CleanShot 2024-02-01 at 14 40 49@2x](https://github.com/logancyang/obsidian-copilot/assets/1957022/e503ddf9-1938-443d-a47c-28768b89ac00)   -- in the screen shot I tried the query once with WIFI on it worked and then again with WIFI off and it didnt!

Thanks in advance.

**Comment by logancyang** [comment]
Hey @borthwick, thanks for reporting! This is entirely on me, I wasn't being clear about local use: currently you still need to have internet connection for QA mode, because the QA needs an embedding model to work, and there's no good local embedding provider solution at the moment. One solution we used to have in Copilot was "LocalAI", but it's quite involved to set up for the average user. I'm actively looking into a local embedding solution right now and will publish a release as soon as I find one.

In the meantime, you can use Chat mode with the small "upload" button to send your notes directly to the prompt. As long as you run a local model with a long context window, it's completely offline and it actually works better than QA mode for more in-depth questions. Let me know if this clears up the confusion!

**Comment by STOI123** [comment]
Still got the same error, the notification info should be clearer... Fetch Error mean nothing to the user

**Comment by YorkyPoo** [comment]
I'm having the same issue.

**Comment by vkichu** [comment]
I am having the same issue, my internet is turned on, I am running ollama from a WSL2 Ubuntu Host and using Obsdian from windows to connect, I see the chat request going to Ollama server log but response code is 403 , expecting 200 so not sure what the issue is.

In the installation for local use there is a comment "Remember that you MUST set this parameter for Ollama models, or they will silently fail and you will think your long prompt successfully reached the model!"  but not sure if this is related to this setting for Ollama models causing the error. If yes can you guide on steps to set the parameters from the Unix prompt, sorry for a newbie question.

**Reply by PT123123** [reply]
Mostly CROS problem causing the Failed to fetch ; As it says in CTRL+SHIFT+I console. I have help some solve the same problem, something about OLLAMA_ORIGINS in env var. Try searching these keywords for more detail.

**Reply by Nielius** [reply]
The crucial part is that you need to run ollama with the right environment variable: if you start with

```bash
OLLAMA_ORIGINS=app://obsidian.md* ollama serve
```

then it should be right.

For me, I'm using systemd to run ollama (you can use `systemctl status ollama` to see if you are also running ollama with systemd), so to add the environment variable there I followed [these instructions](https://serverfault.com/a/413408).

**Reply by Pathsis** [reply]
@Nielius Hello, I'm running ollama on Ubuntu, supposedly as a service through systemd. However, when I run `OLLAMA_ORIGINS=app://obsidian.md* ollama serve`, I always get `Error: listen tcp 127.0.0.1:11434: bind: address already in use`, and when I run `systemctl status ollama`, I get:

```shell
~$ systemctl status ollama
● ollama.service - Ollama Service
     Loaded: loaded (/etc/systemd/system/ollama.service; enabled; preset: enabl>
     Active: active (running) since Wed 2024-05-01 00:31:43 CST; 22h ago
   Main PID: 13879 (ollama)
      Tasks: 15 (limit: 38137)
     Memory: 4.8G (peak: 6.0G)
        CPU: 17min 37.602s
     CGroup: /system.slice/ollama.service
             └─13879 /usr/local/bin/ollama serve

5月 01 12:54:43 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 21:58:04 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 21:58:06 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 21:58:07 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 22:32:02 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 22:32:03 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 22:32:03 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 22:32:59 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 22:33:07 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
5月 01 22:33:40 midtail-Surface-Laptop-Studio ollama[13879]: [GIN] 2024/05/01 ->
```

How should I set it up? Thank you very much!

**Reply by Nielius** [reply]
Hey @Pathsis , the reason you're getting that error is that the systemctl service is already using that port, and when you run `ollama serve`, you're trying to start another service that tries to use that port.

Two solutions:

1. Stop the ollama systemctl service with `systemctl ollama stop`. Then you should be able to run `OLLAMA_ORIGINS=... ollama serve`.
2. Update the ollama systemctl service following [these instructions](https://serverfault.com/a/413408) that I linked to in my previous post, and then doing `systemctl restart ollama` to load the new settings.

**Reply by proofer** [reply]
I'm having the same issue, `Error: listen tcp 127.0.0.1:11434: bind: address already in use`, on MacOS. MacOS has no `systemctl`. `launchctl list | grep ollama` returns nothing. `ps  -x | grep ollama` shows four processes whose binaries are within Ollama.app, including `ollama serve`. I tried killing that process so I could restart it after issuing the assignment to the environment variable, but it is instantly restarted with a different pid.
